From 7963864052ebef675c45072bde65f550fdc23f3a Mon Sep 17 00:00:00 2001
From: Claudiu Manoil <claudiu.manoil@nxp.com>
Date: Wed, 7 Nov 2018 18:35:55 +0200
Subject: [PATCH 472/706] enetc: Refactor default rs tbl setup, avoid zero size
 alloc

Signed-off-by: Claudiu Manoil <claudiu.manoil@nxp.com>
(cherry picked from commit d91bc1387b6fda5eec6cc815f89a008eae3a5f04)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c | 31 +++++++++++++-------
 1 file changed, 21 insertions(+), 10 deletions(-)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index 00f831fe8e6d..e4db0a0e0602 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -967,10 +967,8 @@ static void enetc_clear_cbdr(struct enetc_hw *hw)
 	enetc_wr(hw, ENETC_SICBDRMR, 0);
 }
 
-static int enetc_configure_si(struct enetc_ndev_priv *priv)
+static int enetc_setup_default_rss_table(struct enetc_si *si, int num_groups)
 {
-	struct enetc_si *si = priv->si;
-	struct enetc_hw *hw = &si->hw;
 	int *rss_table;
 	int i;
 
@@ -978,6 +976,23 @@ static int enetc_configure_si(struct enetc_ndev_priv *priv)
 	if (!rss_table)
 		return -ENOMEM;
 
+	/* Set up RSS table defaults */
+	for (i = 0; i < si->num_rss; i++)
+		rss_table[i] = i % num_groups;
+
+	enetc_set_rss_table(si, rss_table, si->num_rss);
+
+	kfree(rss_table);
+
+	return 0;
+}
+
+static int enetc_configure_si(struct enetc_ndev_priv *priv)
+{
+	struct enetc_si *si = priv->si;
+	struct enetc_hw *hw = &si->hw;
+	int err;
+
 	enetc_setup_cbdr(hw, &si->cbd_ring);
 	/* set SI cache attributes */
 	enetc_wr(hw, ENETC_SICAR0,
@@ -987,15 +1002,11 @@ static int enetc_configure_si(struct enetc_ndev_priv *priv)
 	enetc_wr(hw, ENETC_SIMR, ENETC_SIMR_EN);
 
 	if (si->num_rss) {
-		/* Set up RSS table defaults */
-		for (i = 0; i < si->num_rss; i++)
-			rss_table[i] = i % priv->num_rx_rings;
-
-		enetc_set_rss_table(si, rss_table, si->num_rss);
+		err = enetc_setup_default_rss_table(si, priv->num_rx_rings);
+		if (err)
+			return err;
 	}
 
-	kfree(rss_table);
-
 	return 0;
 }
 
-- 
2.17.1

